### ISSUE TITLE
Handling Relationships

### ISSUE THREAD

**Issue by CraigSiemens** [JohnSundell/Unbox#7]
I'm running into an error unboxing related objects that are not nested in the json but associated by id. Here's what my setup looks like.

My json looks like this (except with more objects)

``` json
{
    "vehicles": [
        {
            "type": "Car",
            "color_id": 1
        }
    ],
    "colors": [
        {
            "id": 1,
            "hex": "69AFFC"
        }
    ]
}
```

`Vehicle` and `Color` both conform to `Unboxable`.
`Color` also conforms to `UnboxableByTransform` so that `Vehicle` can get the related color by calling
```self.color = unboxed.unbox("color_id")```

Since `Color` confirms to both protocols, it causes the following compiler error since it can't figure out which `unbox` method to use.

```
Vehicle.swift:53:21: error: ambiguous use of 'unbox'
        self.color = unboxer.unbox("color_id")
                     ^
Unbox.Unboxer:31:17: note: found this candidate
    public func unbox<T : Unboxable>(key: String) -> T?
                ^
Unbox.Unboxer:43:17: note: found this candidate
    public func unbox<T : UnboxableByTransform where T == T.UnboxTransformerType.TransformedType>(key: String) -> T?
                ^
```

Any suggestions for how I should do this instead? My only idea would be to remove `UnboxableByTransform` conformance from `Color` and do something like

```
let colorId: Int = unboxed.unbox("color_id")
self.color = Color.findById(colorId)
```

Still pretty simple but 1 line would be nicer than two :smile:

**Comment by JohnSundell** [comment]
You're right, a type can't conform to both `Unboxable` and `UnboxableByTransform` at the same time. `UnboxableByTransform` is mostly meant for simple value types that can be transformed from a single piece of data. Should probably add more thorough docs about this :)

So there's a few solutions you could potentially use;

You could either make sure the nested `Color` JSON contains enough information to create a `Color` instance, like this:

```
{
    "vehicles": [
        {
            "type": "Car",
            "color": {
               "id": 1,
               "hex": "69AFFC"
            }
        }
    ]
}
```

That way you can unbox the `color` property simply with `unboxer.unbox("color")`, but depending on how the rest of your model code works, it might give you duplicate information.

Another option is to have `Color` simply be a single value type, and only conform to `UnboxableByTransform` and not to `Unboxable`. In case it doesn't need any additonal data from the JSON that might work.

Finally, you have the option of using the `context` API to pass an arbritrary object into the unboxing process. You could, for example, create a `[Int : Color]` dictionary (by unboxing) and then pass it to `Unbox(dictionary, context: colorDictionary)`. The dictionary will then be available using `unboxer.context as? [Int : Color]`.

Hope one of the above solutions work for you! Let me know :)

**Comment by rob-nash** [comment]
I don't think this needs to be the responsibility of the unbox library itself, but rather the author of the json. If you don't have control over the structure of the json, then simply unbox it and map the relationships afterwards.

If the json doesn't have a relationship clearly defined, like so

```
{
    "vehicles": [
        {
            "type": "Car",
            "color": {
               "id": 1,
               "hex": "69AFFC"
            }
        }
    ]
}
```

then the onus is on the user of the library to make that connection.

```
let whatever = ["" : ""]
let vehicle : Vehicle? = Unbox(whatever)
var color : Color? = Unbox(whatever)
vehicle?.color = color
```

**Comment by rob-nash** [comment]
Transformable is only for changing an int to an enum or a string to a url.

**Comment by CraigSiemens** [comment]
Thanks for the feedback.

Unfortunately, I have no control over the structure of the json. In reality it is more complex but I simplified it for example sake.

I have come across other APIs that return relationships this way. It's usually to prevent duplication or because the objects are returned by separate API requests.

I'll probably end up sticking with the solution I came up with. Thanks again.
